interface/TausProducer.h

```diff
@@ -0,0 +1,151 @@
+        TausProducer(const std::string& name, const ROOT::TreeGroup& tree, const edm::ParameterSet& config):
+            LeptonsProducer(name, tree, config), ScaleFactors(const_cast<ROOT::TreeGroup&>(tree))
+        {
+            ScaleFactors::create_branches(config);
+        }
+
+        virtual ~TausProducer() {}
+
+        virtual void doConsumes(const edm::ParameterSet& config, edm::ConsumesCollector&& collector) override {
+            LeptonsProducer::doConsumes(config, std::forward<edm::ConsumesCollector>(collector));
+
+            m_vertices_token = collector.consumes<std::vector<reco::Vertex>>(config.getUntrackedParameter<edm::InputTag>("vertices", edm::InputTag("offlineSlimmedPrimaryVertices")));
+	    m_pruned_token = collector.consumes<std::vector<reco::GenParticle>>(config.getUntrackedParameter<edm::InputTag>("pruned_gen_particles", edm::InputTag("prunedGenParticles")));
+	}
+
+        virtual void produce(edm::Event& event, const edm::EventSetup& eventSetup) override;
+
+	
+    private:
+        // Tokens
+        edm::EDGetTokenT<std::vector<reco::Vertex>> m_vertices_token;
+	edm::EDGetTokenT<std::vector<reco::GenParticle>> m_pruned_token;
+    private:
+ 
+	MatchResult LeptonGenMatch(const LorentzVector& p4, const std::vector<reco::GenParticle>& genParticles);
+
+	inline void FindFinalStateDaughters(const reco::GenParticle& particle, std::vector<const reco::GenParticle*>& daughters,
+					    const std::set<int>& pdg_to_exclude = {})
+	{
+	  if(!particle.daughterRefVector().size()) {
+	    const int abs_pdg = std::abs(particle.pdgId());
+	    if(!pdg_to_exclude.count(abs_pdg))
+	      daughters.push_back(&particle);
+	  } else {
+	    for(const auto& daughter : particle.daughterRefVector())
+	      FindFinalStateDaughters(*daughter, daughters, pdg_to_exclude);
+	  }
+	}
+
+	inline LorentzVector GetFinalStateMomentum(const reco::GenParticle& particle, bool excludeInvisible,
+						   bool excludeLightLeptons)
+	{
+	  using set = std::set<int>;
+	  using pair = std::pair<bool, bool>;
+	  static const set empty = {};
+	  static const set light_leptons = { 11, 13 };
+	  static const set invisible_particles = { 12, 14, 16 };
+	  static const set light_and_invisible = tools::union_sets({light_leptons, invisible_particles});
+
+	  static const std::map<pair, const set*> to_exclude {
+	    { pair(false, false), &empty }, { pair(true, false), &invisible_particles },
+					      { pair(false, true), &light_leptons }, { pair(true, true), &light_and_invisible },
+										       };
+
+	  std::vector<const reco::GenParticle*> daughters;
+	  FindFinalStateDaughters(particle, daughters, *to_exclude.at(pair(excludeInvisible, false)));
+
+	  LorentzVector p4;
+	  for(auto daughter : daughters){
+	    if(excludeLightLeptons && light_leptons.count(std::abs(daughter->pdgId())) && daughter->statusFlags().isDirectTauDecayProduct()) continue;
+	    p4 += daughter->p4();
+	  }
+	  return p4;
+	}
+
+
+
+    public:
+        // Tree members
```

Comment:
Do you need this commented code? If it's optional you can pass a boolean option to TausProducer, if it's not needed there's no reason to keep it around.